### PR TITLE
synchronize: add support for the 'lxd' transport

### PR DESCRIPTION
**Pull request by ghost**
##### SUMMARY

This is done in a similar way than what is done for the `docker`
transport. I had, however, to wrap the `lxc exec` call that is used for
`--rsh` into a temporary bash script. I couldn't manage to do otherwise.

I've also reorganized transport-specific logic a little bit and created
`TransportAdapter` classes. I wanted to avoid the code becoming too
messy by having bits and pieces of transport-specific code everywhere.

I also couldn't manage to make this work without `sudo`. The `su`
version of this fix just wasn't working. Under LXD, containers are
usually very barebone, so they usually don't have `sudo`, so I've added
a check.

I haven't tested the `synchronize` action under docker to make sure it
still works, my current machine isn't geared for it, but I think that my
code changes were rather prudent and I've left the initial logic pretty
much intact.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
synchonize

##### ANSIBLE VERSION
```
ansible-playbook 2.4.0 (synchronize-lxd-transport 7d6eaf33ec) last updated 2017/07/17 16:21:27 (GMT -400)
  config file = None
  configured module search path = [u'/home/vdupras/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vdupras/src/ansible/lib/ansible
  executable location = /home/vdupras/.local/bin/ansible-playbook
  python version = 2.7.12 (default, Jun  9 2017, 11:18:57) [GCC 5.4.0]
```

### CODE REVIEW

**Comment by ghost** [comment]
Sorry for the test failures. I was on my laptop which wasn't fully configured for ansible's test suite and I didn't run them. I'll fix this soon.

**Comment by hsoft** [comment]
Yup, it looks like I broke synchronize + docker. will fix soon. Flagging as WIP until then.

**Comment by ghost** [comment]
Fixed. There are still failures, but they're the same as in #26930, which is a trivial change, so I'm guessing that failures aren't related to the change.

**Comment by jctanner** [comment]
bot_status

**Comment by erwilan** [comment]
please consider #24911

**Comment by ghost** [comment]
I have rebased the PR and chipped in an integration test! as I write in #27437, I wasn't sure of how to do it elegantly and I tried many things. This tests follows @mattclay's suggestions.

**Comment by ghost** [comment]
I've also added a `synchronize_docker` integration test so that we can easily verify that this PR doesn't break the docker part.

File: test/integration/targets/synchronize_docker/aliases [hunk 1]
```diff
@@ -0,0 +1 @@
+non_local
```
Comment (mattclay) [inline test/integration/targets/synchronize_docker/aliases:1]: This test should have `posix/ci/group1` for an alias so the tests run in CI.

Comment (ghost) [reply test/integration/targets/synchronize_docker/aliases:1]: Can the CI run a test depending on a docker transport at all? [connection_docker](https://github.com/ansible/ansible/blob/devel/test/integration/targets/connection_docker/aliases) isn't tagged for running on the CI either and it has the same requirements.

**Comment by ghost** [comment]
Oh well, let's close this, it's never going to be merged.

**Pull request closed without merging**